software/py/vanilla_stats_parser.py

```diff
@@ -0,0 +1,640 @@
+                         "stall_data"      : type_format["name"] + type_format["int"]  + type_format["percent"] + type_format["percent"] + "\n",
+                         "miss_header"     : type_format["name"] + type_format["type"] + type_format["type"]    + type_format["type"]    + "\n",
+                         "miss_data"       : type_format["name"] + type_format["int"]  + type_format["int"]     + type_format["float"]   + "\n",
+                         "line_break"      : '=' * 90 + "\n"
+                        }
+
+
+
+    # list of instructions, operations and events parsed from vanilla_stats.log
+    # populated by reading the header of input file 
+    self.stats_list   = []
+    self.instr_list   = []
+    self.miss_list    = []
+    self.stalls_list  = []
+    self.all_ops_list = []
+
+    # Call generate_stats after initialization
+    self.generate_stats(input_file)
+    return
+
+
+
+
+  # print a line of stat into stats file based on stat type
+  def __print_stat(self, stat_file, stat_type, *argv):
+    stat_file.write(self.print_format[stat_type].format(*argv));
+    return
+
+
+  # Decodes the tag value of the stat to determine the type of 
+  # stat, the tile group id of sending tile, and the tag value 
+  def __decode_stat_val(self, stat_val):
+    stat_type   = (stat_val >> (bsg_STAT_TAG_BITS + bsg_STAT_TG_ID_BITS + bsg_STAT_X_BITS + bsg_STAT_Y_BITS))
+    stat_y      = (stat_val >> (bsg_STAT_TAG_BITS + bsg_STAT_TG_ID_BITS + bsg_STAT_X_BITS)) & bsg_STAT_Y_MASK
+    stat_x      = (stat_val >> (bsg_STAT_TAG_BITS + bsg_STAT_TG_ID_BITS)) & bsg_STAT_X_MASK
+    stat_tg_id  = (stat_val >> (bsg_STAT_TAG_BITS)) & bsg_STAT_TG_ID_MASK
+    stat_tag    = (stat_val) & bsg_STAT_TAG_MASK
+    return (stat_type, stat_y, stat_x, stat_tg_id, stat_tag)
+    
+
+
+  # go though the input traces and extract start and end stats  
+  # for each tile, and each tile group 
+  # return number of tile groups, tile group timing stats, and the tile stats
+  # this function only counts the portion between two print_stat_start and end messages
+  # in theory, this excludes the time tiles are waiting to be loaded, etc.
+  def __generate_tile_stats(self, traces):
+
+    num_tile_groups = 0
+
+    tile_stat_start = [[Counter() for x in range(self.manycore_dim_x)] for y in range(self.manycore_dim_y)]
+    tile_stat_end   = [[Counter() for x in range(self.manycore_dim_x)] for y in range(self.manycore_dim_y)]
+    tile_stat       = [[Counter() for x in range(self.manycore_dim_x)] for y in range(self.manycore_dim_y)]
+
+    tile_group_stat_start = [Counter() for tg_id in range(self.max_tile_groups)] 
+    tile_group_stat_end   = [Counter() for tg_id in range(self.max_tile_groups)] 
+    tile_group_stat       = [Counter() for tg_id in range(self.max_tile_groups)] 
+
+    for trace in traces:
+      y = trace["y"]
+      x = trace["x"]
+      relative_y = y - bsg_ORG_Y
```

Comment:
Is it worth worrying about relative x/y coordinates? Why not just print absolute X/Y coordinates?

Reply:
I need relative because of indexing into an array. If dimensions are 4x4, I'm instantiating a 4x4 array, but if the absolute X/Y is between 1 to 5, then it goes out of index.